src/lib/drivers/smbus/SMBus.cpp

```diff
@@ -0,0 +1,163 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "SMBus.hpp"
+
+SMBus::SMBus(int bus_num, uint16_t address) :
+	I2C("BATT_SMBUS_I2C", nullptr, bus_num, address, 100000)
+{
+	init();
+}
+
+SMBus::~SMBus()
+{
+}
+
+int SMBus::read_word(const uint8_t cmd_code, void *data)
+{
+	// 2 data bytes + pec byte
+	int result = transfer(&cmd_code, 1, (uint8_t *)data, 3);
+
+	if (result == PX4_OK) {
+		// Check PEC.
+		uint8_t addr = get_device_address() << 1;
+		uint8_t full_data_packet[5];
+		full_data_packet[0] = addr | 0x00;
+		full_data_packet[1] = cmd_code;
+		full_data_packet[2] = addr | 0x01;
+		memcpy(&full_data_packet[3], data, 2);
+
+		uint8_t pec = get_pec(full_data_packet, sizeof(full_data_packet) / sizeof(full_data_packet[0]));
+
+		if (pec == ((uint8_t *)data)[2]) {
+			return PX4_OK;
+
+		} else {
+			return PX4_ERROR;
+		}
+	}
+
+	return PX4_ERROR;
+}
+
+int SMBus::block_read(const uint8_t cmd_code, void *data, const uint8_t length)
+{
+	unsigned byte_count = 0;
+	// Length of data (32max). byte_count(1), cmd_code(2), pec(1)
+	uint8_t rx_data[32 + 4];
+
+	transfer(&cmd_code, 1, (uint8_t *)rx_data, length + 2);
+
+	byte_count = rx_data[0];
+
+	// addr1, addr2, byte_count,
+	memcpy(data, &rx_data[1], byte_count);
+
+	// addr(wr), cmd_code, addr(r), byte_count, rx_data[]
+	uint8_t device_address = get_device_address();
+	uint8_t full_data_packet[32 + 4] = {0};
+
+	full_data_packet[0] = (device_address << 1) | 0x00;
+	full_data_packet[1] = cmd_code;
+	full_data_packet[2] = (device_address << 1) | 0x01;
+	full_data_packet[3] = byte_count;
+
+	memcpy(&full_data_packet[4], &rx_data[1], byte_count);
```

Comment:
We can avoid this memcpy if `transfer` copies the data directly into `full_data_packet` at the proper offset.

Reply:
The PEC does not use the first byte returned from `transfer` (the byte count)